stdlib/public/core/Result.swift

```diff
@@ -0,0 +1,131 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A value that represents either a success or failure, capturing associated
+/// values in both cases.
+@_frozen
+public enum Result<Success, Failure: Error> {
+  /// A success, storing a `Success` value.
+  case success(Success)
+  
+  /// A failure, storing a `Failure` value.
+  case failure(Failure)
+  
+  /// Evaluates the given transform closure when this `Result` instance is
+  /// `.success`, passing the value as a parameter.
+  ///
+  /// Use the `map` method with a closure that returns a non-`Result` value.
+  ///
+  /// - Parameter transform: A closure that takes the successful value of the
+  ///   instance.
+  /// - Returns: A new `Result` instance with the result of the transform, if
+  ///   it was applied.
+  public func map<NewSuccess>(
+    _ transform: (Success) -> NewSuccess
```

Comment:
I don't remember it mentioned in the forum discussions, but did we ever consider accepting throwing closures here?

Reply:
Yes, there was some discussion in the initial pitch as well as earlier reviews. Some expressed discomfort around manipulating an already error-containing value with transforms that could throw, and what that would mean. Previous versions of this PR had an implementation of a `flatMap` overload that took a throwing closure that, rather than rethrowing, captured the error as part of `Result` output. This was removed by the core team. I think such transforms are useful, but the core team would like better naming.